src/polyswarm_api/endpoint.py

```diff
@@ -0,0 +1,254 @@
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES, request_cls=None):
+        self.session = request_cls(key, retries)
+        self.unauth_session = request_cls(key=None, retries=retries)
+        self.timeout = timeout
+
+    def execute(self, request):
+        if 'timeout' not in request:
+            request['timeout'] = self.timeout
+        return self.session.request(**request)
+
+    def unauth_execute(self, request):
+        return self.unauth_session.request(**request)
+
+    def _get_engine_names(self, request):
+        return self.unauth_session.request(**request)
+
+    def __getattr__(self, name):
+        if name in self.__dict__:
+            return self.__dict__[name]
+        return self.execute
+
+    def _download_to_fh(self, req, fh):
+        raise NotImplementedError
+
+    def download(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.execute(request)
+        return self._download_to_fh(req, fh)
+
+    def download_archive(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.unauth_execute(request)
+        return self._download_to_fh(req, fh)
+
+
+class PolyswarmFuturesExecutor(PolyswarmRequestExecutor):
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES):
+        super(PolyswarmFuturesExecutor, self).__init__(key, timeout, retries, PolyswarmHTTPFutures)
+
+    def _download_to_fh(self, req, fh):
+        # this is unfortunately the cleanest way I think I can do this with requests-futures
+        # derived partially from https://github.com/ross/requests-futures/issues/54
+        def do_download(r, f):
+            for chunk in r.iter_content(chunk_size=const.DOWNLOAD_CHUNK_SIZE):
+                f.write(chunk)
+            return r
+        resp = req.result()
+        resp.raise_for_status()
+        return self.session.executor.submit(do_download, resp, fh)
+
+
+class PolyswarmSynchronousExecutor(PolyswarmRequestExecutor):
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES):
+        super(PolyswarmSynchronousExecutor, self).__init__(key, timeout, retries, PolyswarmHTTP)
+
+    def _download_to_fh(self, req, fh):
+        for chunk in req.iter_content(chunk_size=const.DOWNLOAD_CHUNK_SIZE):
+            fh.write(chunk)
+        return req
+
+
+class PolyswarmEndpoint(object):
+    """ This is the base class for PolyswarmEndpoint classes. Do not use directly. """
+    def __init__(self,  key, uri=const.DEFAULT_GLOBAL_API, community=const.DEFAULT_COMMUNITY,
+                 timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES,
+                 request_gen_cls=PolyswarmRequestGenerator, request_exec_cls=PolyswarmFuturesExecutor):
```

Comment:
I would not get the `class` here, but rather an `instance` already created. The reason for that is that a user might want to create their own request generator and executor that does not get initialized as you are expecting. We are making assumptions about how these classes should be initialized and not making use of `duck typing` here. We don't really care how to initialize these classes, we only care that they implement the correct methods. It is a better pattern to let the caller initialize these. My suggestion would be:

```python
    def __init__(self,  key, uri=const.DEFAULT_GLOBAL_API, community=const.DEFAULT_COMMUNITY,
                 timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES,
                 generator=None, executor=None):
        self.executor = executor or PolyswarmFuturesExecutor(key, timeout, retries)
        self.generator = generator or PolyswarmRequestGenerator(uri, community)
```

We still provide sane defaults while allowing the user to implement their own classes. This is a more flexible pattern.

I would also make the default the simplest executor (even though it might not be the most optimized). I would always try to keep things simple. `async` stuff, dealing with futures and multiprocessing is an optimization, and should be considered advanced stuff that adds significantly complexity. If I run a method, I would like to get a response right away rather than a future. To quote Don Knuth, "premature optimization is the root of all evil", that is why I would rather let them optimize only if they need it.

Reply:
You can just say you don't like `async`, no need to quote Knuth :P

Fine by me.